docs(api): new_pipette.rst revisions for Flex

jwwojak opened the pull request:
# Overview

Suggested text and code example revisions to the [Pipette](https://docs.opentrons.com/v2/new_pipette.html) API docs (new_pipette.rst file) to include Flex text and code examples.

Sandbox link: http://sandbox.docs.opentrons.com/pipette-docs-revisions/v2/

# Test Plan

- Checked links and cross-references to make sure they work.
- Asking reviewers to check doc code changes. Revisions replace sample code showing OT-2 examples in favor of Flex examples.   

# Changelog

Changes include:

- Extensive text re-writing throughout.
- Some new headers and reorganization.
- Removed some references or text to OT-2 GEN1 pipettes. Emphasis is on OT-2 GEN2 and Flex pipettes.
- Sample code changes from OT-2 examples to Flex.
- Simplifying flow rate data. Remove duplicated info, use tables for better info display.

# Review requests

Please 2x check the code samples.

Also, perhaps @SyntaxColoring could take a look at the section titled "Loading a Flex 96-Channel Pipette." The example puts the 96 channel in the left slot( `mount='left'`). It's a simple `load_instrument` example, but maybe there's more to add or a better example to give users. Any input would be appreciated.

# Risk assessment

Low. Documentation changes only. Any removed OT-2 GEN1 pipette info has been saved separately in case we need to restore it.

jwwojak committed:
Starting on pipettes

Mid-day save

jwwojak committed:
Update new_pipette.rst

jwwojak committed:
Update new_pipette.rst

End of day commit

jwwojak committed:
Merge branch 'edge' into pipette-docs-revisions

jwwojak committed:
Update new_pipette.rst

Midmorning save, just in case.

jwwojak committed:
Update new_pipette.rst

Culling tables and organizing pipette types. Lunch commit.

jwwojak committed:
Update new_pipette.rst

End of day save

jwwojak committed:
Merge branch 'edge' into pipette-docs-revisions

jwwojak committed:
Update new_pipette.rst

Mid day save.

jwwojak committed:
Update new_pipette.rst

End of day save.

jwwojak committed:
Merge branch 'edge' into pipette-docs-revisions

jwwojak committed:
Morning update

Early morning save up.

jwwojak committed:
Update new_pipette.rst

Moar saving.

jwwojak committed:
Update new_pipette.rst

More before lunch.

jwwojak committed:
Update new_pipette.rst

mid afternoon save

jwwojak committed:
Update new_pipette.rst

End of day revisions and save.

jwwojak committed:
Merge branch 'edge' into pipette-docs-revisions

jwwojak committed:
Update new_pipette.rst

Top 'o the morning save.

jwwojak committed:
Update new_pipette.rst

Updated flow rates. May open draft PR.

jwwojak committed:
Update new_pipette.rst

Added comment about table alignment.

jwwojak committed:
Update new_pipette.rst

Small tweak. final final final. Open pr.

jwwojak committed:
Update new_pipette.rst

Fix/condense flow rate tables.

jwwojak committed:
Update new_pipette.rst

Change flow rate section title. Add some comments.

jwwojak committed:
Update new_pipette.rst

Change header for GEN2 - GEN1 pipette comparisons

jwwojak committed:
Update new_pipette.rst

Table formatting fix because of comment on preceeding line.

jwwojak committed:
Update new_pipette.rst

Remove internal comment. Fix flex flow rate text to match ot-2 text.

ecormany commented on api/docs/v2/new_pipette.rst:
```diff
@@ -6,78 +6,119 @@
 
-.. code-block:: python
-    :substitutions:
+Similar to working with labware and modules, you must inform the robot about the pipettes you want to use in your protocol. Even if you don't use the pipette anywhere else in your protocol, the Opentrons App and the robot won't let you start the protocol run until all pipettes loaded by ``load_instrument()`` are attached properly.
```
this language, "you must inform the robot about the pipettes you want to use in your protocol", repeats the opening sentence too closely. the rest of the info in this ¶ is good.

jwwojak replied:
Removing that paragraph and incorporating into 2nd paragraph of 1st section. 
 
> Similar to working with labware and modules, you must inform the robot about the pipettes you want to use in your protocol. As noted above, the:py:meth:`~.ProtocolContext.load_instrument` method provides this capability. It also requires the :ref:`pipette's API load name <new-pipette-models>`, its left or right mount position, and (optionally) a list of associated tip racks. Even if you don't use the pipette anywhere else in your protocol, the Opentrons App and the robot won't let you start the protocol run until all pipettes loaded by ``load_instrument()`` are attached properly.

ecormany commented on api/docs/v2/new_pipette.rst:
```diff
@@ -6,78 +6,119 @@
+Loading Flex 1- and 8-Channel Pipettes
+--------------------------------------
 
+This code sample loads a Flex 8-Channel Pipette in the left slot along with a Flex 1-Channel Pipette and two tip racks in the right slot. Both pipettes are 1000 µL. 
```
makes it sound like the tip racks are in the right pipette mount! either remove that detail, or move it to a second sentence.

also, to avoid confusion, always use _mount_ for pipette locations. _slots_ are on the deck.

jwwojak replied:
Replacing "slot" with "mount."

ecormany commented on api/docs/v2/new_pipette.rst:
```diff
@@ -6,78 +6,119 @@
+    requirements = {'robotType': 'Flex', 'apiLevel': '|apiLevel|'}
+
+    def run(protocol: protocol_api.ProtocolContext):
+        left = protocol.load_instrument(
+            instrument_name='flex_8channel_1000', mount='left')   
+        tiprack1 = protocol.load_labware(
+            load_name='opentrons_flex_96_tiprack_1000ul', location='D1')
+        tiprack2 = protocol.load_labware(
+            load_name='opentrons_flex_96_tiprack_1000ul', location='C1')
+        right = protocol.load_instrument(
+            instrument_name='flex_1channel_1000',
+            mount='right',
+            tip_racks=[tiprack1, tiprack2]) 
```
it's a little weird to assign no tip racks to one pipette and two to the other — this isn't best practice, so it's a little weird to neutrally present it as the first example of `load_instrument()`. i think it would be far better to have 1 rack per pipette.

jwwojak replied:
Code samples from original, OT-2 version. Revising to use 1 rack per pipette, changing to load 1 one channel and 8 channel.

ecormany commented on api/docs/v2/new_pipette.rst:
```diff
@@ -6,78 +6,119 @@
+            instrument_name='flex_96channel_1000', mount='left')
 
-.. versionadded:: 2.0
+.. version added, 2.15??
```
yes, that seems right. `flex_96channel_1000` is a valid load name in 2.15.

jwwojak replied:
Added version.

ecormany commented on api/docs/v2/new_pipette.rst:
```diff
@@ -6,78 +6,119 @@
+---------------------
 
-    When you load a pipette in a protocol, you inform the OT-2 that you want the specified pipette to be present. Even if you do not use the pipette anywhere else in your protocol, the Opentrons App and the OT-2 will not let your protocol proceed until all pipettes loaded with ``load_instrument`` are attached to the OT-2.
+This code sample loads a 8-channel, P300 GEN2 pipette in the left slot and a single-channel, P1000 GEN2 pipette with two tip racks in the right slot. 
```
use standard product name order:
- P300 8-Channel GEN2
- P100 Single-Channel GEN2

jwwojak replied:
Replacing with correct product names.

- Flex 1-Channel Pipette
- Flex 8-Channel Pipette

ecormany commented on api/docs/v2/new_pipette.rst:
```diff
@@ -6,78 +6,119 @@
+
+    from opentrons import protocol_api
+
+    metadata = {'apiLevel': '2.14'}
```
nothing stopping this from being `|apiLevel|` that i know of

jwwojak replied:
Revised as:

> metadata = {'apiLevel': |apiLevel|}

ecormany commented on api/docs/v2/new_pipette.rst:
```diff
@@ -6,78 +6,119 @@
+    metadata = {'apiLevel': '2.14'}
+
+    def run(protocol: protocol_api.ProtocolContext):
+        left = protocol.load_instrument(
+            instrument_name='p300_multi_gen2',
+            mount='left')
+        tiprack1 = protocol.load_labware(
+            load_name='opentrons_96_tiprack_1000ul',
+            location=1)
+        tiprack2 = protocol.load_labware(
+            load_name='opentrons_96_tiprack_1000ul',
+            location=2)
+        right = protocol.load_instrument(
+            instrument_name='p1000_single_gen2',
+            mount='right',
+            tip_racks=[tiprack1, tiprack2])
```
mirror any change to tip rack loading in Flex example above.

jwwojak replied:
These have changed and mirror each other.

- loading tip racks 1st
- loading instruments after

ecormany commented on api/docs/v2/new_pipette.rst:
```diff
@@ -6,78 +6,119 @@
-pipettes will always aspirate and dispense on all channels simultaneously.
-
-For instance, to aspirate from the first column of a 96-well plate you would write:
+Because there is only one motor in a multi-channel pipette, these pipettes always aspirate and dispense on all channels simultaneously. For instance, to aspirate from the first column of a 96-well plate you would write:
 
 .. code-block:: python
```
```suggestion
Because there is only one motor in a multi-channel pipette, these pipettes always aspirate and dispense on all channels simultaneously. For instance, to aspirate from the first column of a 96-well plate::
```
i'm not a big fan of the "you would write" type constructions. i prefer the concision of "To do X: code for X".

jwwojak replied:
The example after this text, I've separated into two H3 sections. Seem to work better as there's a lot of info being conveyed here. New outline is:

- H2 Multi-Channel Pipettes and Well Plates
- H3 8-Channel, 96-Well Plate Example
- H3 8-Channel, 384-Well Plate Example

This was also the big code sample including extensive multi-line comments. Broke it up. Took comments out of code and incorporated into text with short code examples.

ecormany commented on api/docs/v2/new_pipette.rst:
```diff
@@ -6,78 +6,119 @@
     def run(protocol: protocol_api.ProtocolContext):
-        # Load a tiprack for 300uL tips
-        tiprack1 = protocol.load_labware('opentrons_96_tiprack_300ul', 1)
+        # Load a tiprack for 1000uL tips
```
```suggestion
        # Load a tip rack for 1000 µL tips
```
let's clean these up to match house style

ecormany commented on api/docs/v2/new_pipette.rst:
```diff
@@ -6,78 +6,119 @@
+        tiprack1 = protocol.load_labware(
+        load_name='opentrons_flex_96_tiprack_1000ul',
+        location='D1')
         # Load a wellplate
```
```suggestion
        # Load a well plate
```
here too

jwwojak replied:
Fixed throughout.

ecormany commented on api/docs/v2/new_pipette.rst:
```diff
@@ -87,187 +128,169 @@ For instance, to aspirate from the first column of a 96-well plate you would wri
         # Dispense into column 3 of the plate with all 8 channels of the
         # pipette at the top of their respective wells
-        right.dispense(300, plate['A3'].top())
+        right.dispense(volume=300, location=plate['A3'].top())
```
These long code comments are hard to read. I'd rather break up the snippet than make people read dark grey monospace on light grey.

ecormany commented on api/docs/v2/new_pipette.rst:
```diff
@@ -87,187 +128,169 @@ For instance, to aspirate from the first column of a 96-well plate you would wri
-other well starting with the first (rows A, C, E, G, I, K, M, and O); specifying well
-B1 similarly accesses every other well, but starting with the second (rows B, D,
-F, H, J, L, N, and P).
+In general, you should specify wells in the first row of a well plate when using multi-channel pipettes. One exception to this rule is when using 384-well plates. The limited space between the wells in a 384-well plate and between the nozzles of a multi-channel pipette means the pipette accesses every other well in a column. Specifying well A1 accesses every other well starting with the first (ro ...
```
it's not that the space is "limited" — kind of the opposite. 384 plates have double the density (horizontal and vertical) of 96 plates.

jwwojak replied:
Revised to remove "limited."

> The greater well density means the nozzles of a multi-channel pipette can only accesses every other well in a column.

ecormany commented on api/docs/v2/new_pipette.rst:
```diff
@@ -87,187 +128,169 @@ For instance, to aspirate from the first column of a 96-well plate you would wri
+            load_name='corning_384_wellplate_112ul_flat', location=4)
 
-        # Load a P300 Multi GEN2 on the right mount
+        # Load an 8-channel Multi GEN2 on the right mount
```
```suggestion
        # Load a P300 8-Channel GEN2 on the right mount
```

jwwojak replied:
This is fixed in a recent draft push.

ecormany commented on api/docs/v2/new_pipette.rst:
```diff
@@ -87,187 +128,169 @@ For instance, to aspirate from the first column of a 96-well plate you would wri
 
         # Aspirate from wells A1, C1, E1, G1, I1, K1, M1, and O1
-        right.aspirate(300, plate['A1'])
+        right.aspirate(volume=100, location=plate['A1'])
```
good catch! 300 µL doesn't fit in a 384-well plate.

ecormany commented on api/docs/v2/new_pipette.rst:
```diff
@@ -87,187 +128,169 @@ For instance, to aspirate from the first column of a 96-well plate you would wri
+        +-------------------------+-----------+-------------------------+
+        | Pipette Name            | Capacity  | API Load Name           |
+        +=========================+===========+=========================+
+        | Flex 1-Channel Pipette  | 0.5–50 µL | ``flex_1channel_50``    |
+        +                         +-----------+-------------------------+
+        |                         | 5–1000 µL | ``flex_1channel_1000``  |
+        +-------------------------+-----------+-------------------------+
+        | Flex 8-Channel Pipette  + 0.5–50 µL + ``flex_8channel_50``    |
```
min volume of smaller Flex pipettes is 1 µL, not 0.5

jwwojak replied:
Fixing minimum volume in next push.  Change to 1-50 uL

ecormany commented on api/docs/v2/new_pipette.rst:
```diff
@@ -87,187 +128,169 @@ For instance, to aspirate from the first column of a 96-well plate you would wri
+        +-----------------------------+--------------------+-----------------------+
+        | Pipette Name                | Capacity           | API Load Name         |
+        +=============================+====================+=======================+
+        | P20 GEN2 (single channel)   | 1-20 µL            | ``p20_single_gen2``   |
+        +-----------------------------+                    +-----------------------+
+        | P20 GEN2 (8-channel)        |                    | ``p20_multi_gen2``    |
+        +-----------------------------+--------------------+-----------------------+
+        | P300 GEN2 (single chanel)   | 20-300 µL          | ``p300_single_gen2``  |
+        +-----------------------------+                    +-----------------------+
+        | P300 GEN2 (8-channel)       |                    | ``p300_multi_gen2``   |
+        +-----------------------------+--------------------+-----------------------+
+        | P1000 GEN2 (single channel) | 100-1000 µL        | ``p1000_single_gen2`` |
```
reorder, capitalize, and hyphenate names: e.g. P20 Single-Channel GEN2

jwwojak replied:
Fixed to use proper pipette names e.g.

- P20 Single-Channel GEN2
- Flex 1-Channel Pipette

ecormany commented on api/docs/v2/new_pipette.rst:
```diff
@@ -87,187 +128,169 @@ For instance, to aspirate from the first column of a 96-well plate you would wri
+
+The OT-2 works with the GEN1 and GEN2 pipette models. The newer GEN2 pipettes have different volume ranges than the older GEN1 pipettes. With some exceptions, the volume ranges for GEN2 pipettes overlap those used by the GEN1 models. If your protocol specifies a GEN1 pipette, but you have a GEN2 pipette with a compatible volume range, you can still run your protocol. The OT-2 will consider the GE ...
+
+.. list-table::
```
as i read the formatted table, i find that my eye glances right past the volume figures. what if we had three columns: GEN2 Pipette, GEN1 Pipette, GEN1 min. vol.?

the first row then would say P20 Single GEN2, P10 Single GEN1, 1 µL

jwwojak replied:
Fixed to show GEN2 name, GEN1 name, GEN1 volume.

ecormany commented on api/docs/v2/new_pipette.rst:
```diff
@@ -87,187 +128,169 @@ For instance, to aspirate from the first column of a 96-well plate you would wri
-
-    metadata = {'apiLevel': '|apiLevel|'}
-
+The ``load_instrument()`` method includes the optional argument, ``tip_racks``. This parameter accepts a list of tip rack labware objects, which lets you to specify as many tip racks as you want. The advantage of using ``tip_racks`` is twofold. First, associating tip racks with your pipette allows for automatic tip tracking throughout your protocol. Second, it removes the need to specify tip loca ...
```
```suggestion
The ``load_instrument()`` method includes the optional argument ``tip_racks``. This parameter accepts a list of tip rack labware objects, which lets you to specify as many tip racks as you want. The advantage of using ``tip_racks`` is twofold. First, associating tip racks with your pipette allows for automatic tip tracking throughout your protocol. Second, it removes the need to specify tip locations in the :py:meth:`.InstrumentContext.pick_up_tip` method. For example::
```
not an appositive. i will die on this hill 😆

jwwojak replied:
No need to die. Removing comma.

ecormany commented on api/docs/v2/new_pipette.rst:
```diff
@@ -87,187 +128,169 @@ For instance, to aspirate from the first column of a 96-well plate you would wri
+            mount='right',
+            tip_racks=[tiprack_right])
+
+    # You must specify the tip location for the left pipette, which was
```
i'd also like to split these comments out into body paragraphs.

jwwojak replied:
Yes, new changes break up the code and move revised comments to regular text.

ecormany commented on api/docs/v2/new_pipette.rst:
```diff
@@ -87,187 +128,169 @@ For instance, to aspirate from the first column of a 96-well plate you would wri
-Opentrons pipettes aspirate or dispense at different rates. These flow rates can be changed on a loaded
-:py:class:`.InstrumentContext` at any time, in units of µL/sec by altering
-:py:obj:`.InstrumentContext.flow_rate`. This has the following attributes:
+Pipettes aspirate or dispense at different rates measured in units of µL/second. You can change the flow rate on a loaded :py:class:`.InstrumentContext` by altering the :py:obj:`~.InstrumentContext.flow_rate` properties listed below:
```
probably `µL/s` or `microliters per second`. this feels like a hybrid.

jwwojak replied:
Committing in next push.

ecormany commented on api/docs/v2/new_pipette.rst:
```diff
@@ -303,138 +326,46 @@ Each of these attributes can be altered without affecting the others.
         pipette.drop_tip()
 
-
+.. Docs notes this property was removed in 2.14. If removed, why keep?
 :py:obj:`.InstrumentContext.speed` offers the same functionality, but controlled in
 units of mm/s of plunger speed. This does not have a linear transfer to flow rate and
 should only be used if you have a specific need.
```
Docs cover all versions, up through the latest and greatest (2.15). It's totally valid to use older versions on OT-2, even into the future. But this was _removed for a good reason_ so we should probably caution users against it fairly strongly. Something like:

```
.. note::
    In API version 2.13 and earlier, :py:obj:`.InstrumentContext.speed` offered similar functionality. It attempted to set the plunger speed in mm/s. Due to technical limitations, that speed could only be approximate. You must use ``.flow_rate`` in version 2.14 and later, and you should consider replacing older code that sets ``.speed``.
```

jwwojak replied:
Committing in next push.

ecormany commented on api/docs/v2/new_pipette.rst:
```diff
@@ -303,138 +326,46 @@ Each of these attributes can be altered without affecting the others.
+The following table provides data on the default aspirate, dispense, and blow-out flow rates (in µL/s) for OT-2 GEN2 pipettes.
+
++---------------------------------+---------------+---------------+---------------+
+| Pipette                         | Aspirate      | Dispense      | Blow-out      |
```
I don't think we need the split header cell here. But we may want to split volume into its own column, since it comes at the end of every entry under Pipette. So:
Pipette | Volume range | Aspirate, dispense, blow-out rate

ecormany replied:
or just "Flow rate" for the third column

jwwojak replied:
Went with "Flow Rates (uL/s)". Including the units in header for extra reinforcement/reminder.

ecormany commented on api/docs/v2/new_pipette.rst:
```diff
@@ -303,138 +326,46 @@ Each of these attributes can be altered without affecting the others.
+| P300 Multi GEN2 (20-300 µL)     |    94 µL/s                                    |
++---------------------------------+-----------------------------------------------+
+
+.. what is head speed, is it important/significant? applies to gantry movement or pipette? need to find out.
```
based on the following line, i think that's the speed that the gantry moves the pipette around at.

ecormany commented on api/docs/v2/new_pipette.rst:
```diff
@@ -303,138 +326,46 @@ Each of these attributes can be altered without affecting the others.
++---------------------------------+-----------------------------------------------+
+
+.. what is head speed, is it important/significant? applies to gantry movement or pipette? need to find out.
+Additionally, all OT-2 GEN2 pipettes have a default head speed of 400 mm/second and a well bottom clearance of 1mm for aspirate and dispense actions.
```
```suggestion
Additionally, all OT-2 GEN2 pipettes have a default head speed of 400 mm/second and a well bottom clearance of 1 mm for aspirate and dispense actions.
```
🧐

jwwojak replied:
Committing and resolving.

ecormany reviewed (changes requested):
Whew, this page is in fact _a lot_! It's definitely organized better than before. I did leave several comments on improving structure further. Another pass through and this should be good.

ecormany committed:
Merge branch 'edge' into pipette-docs-revisions

jwwojak committed:
Revisons I

Lunch time special. Saving. More tk.

jwwojak committed:
Update new_pipette.rst

Moved tip racks up in the TOC. Added subsections for multi-channel pipettes and well plates. Pushing from local and will now address on-line repo comments.

jwwojak committed:
Update new_pipette.rst

A decent draft push. This is ready for the regular PR process, I think. Will push and convert.

jwwojak commented:
Turning this into a regular PR and re-requesting review.

jwwojak committed:
Remove duplicated text

Duplicated similar text that should have appeared in a Note. Removing the similar text.

jwwojak committed:
Adding Flex flow rate data

And revising some tables and table headers.

ecormany commented on api/docs/v2/new_pipette.rst:
```diff
@@ -15,257 +15,211 @@ For information about liquid handling, see :ref:`v2-atomic-commands` and :ref:`v
-You load pipettes in a protocol using the :py:meth:`~.ProtocolContext.load_instrument` method. This method requires the :ref:`pipette's API load name <new-pipette-models>`, its left or right mount position, and (optionally) a list of associated tip racks.
-
-Similar to working with labware and modules, you must inform the robot about the pipettes you want to use in your protocol. Even if you don't use the pipette anywhere else in your protocol, the Opentrons App and the robot won't let you start the protocol run until all pipettes loaded by ``load_instrument()`` are attached properly.
+Similar to working with labware and modules, you must inform the robot about the pipettes you want to use in your protocol. As noted above, the:py:meth:`~.ProtocolContext.load_instrument` method provides this capability. It also requires the :ref:`pipette's API load name <new-pipette-models>`, its left or right mount position, and (optionally) a list of associated tip racks. Even if you don't use ...
```
```suggestion
Similar to working with labware and modules, you must inform the robot about the pipettes you want to use in your protocol. As noted above, the :py:meth:`~.ProtocolContext.load_instrument` method provides this capability. It also requires the :ref:`pipette's API load name <new-pipette-models>`, its left or right mount position, and (optionally) a list of associated tip racks. Even if you don't use the pipette anywhere else in your protocol, the Opentrons App and the robot won't let you start the protocol run until all pipettes loaded by ``load_instrument()`` are attached properly.
```
syntax fix

ecormany commented on api/docs/v2/new_pipette.rst:
```diff
@@ -15,257 +15,211 @@ For information about liquid handling, see :ref:`v2-atomic-commands` and :ref:`v
 
 .. versionadded:: 2.0
 
+When you load a pipette in this way, you are declaring that you want the specified pipette to be attached to the robot. Even if you don't use the pipette anywhere else in your protocol, the Opentrons App or the touchscreen on Flex will not let your protocol proceed until all pipettes loaded with ``load_instrument`` are attached.
```
can delete, this is covered in the intro.

jwwojak replied:
oops. fat finger. Fixing this locally. Will change and push.

ecormany commented on api/docs/v2/new_pipette.rst:
```diff
@@ -15,257 +15,211 @@ For information about liquid handling, see :ref:`v2-atomic-commands` and :ref:`v
 
+When you load a pipette in this way, you are declaring that you want the specified pipette to be attached to the robot. Even if you don't use the pipette anywhere else in your protocol, the Opentrons App or the touchscreen on Flex will not let your protocol proceed until all pipettes loaded with ``load_instrument`` are attached.
+
+If you're writing a protocol that uses the Flex Gripper, you might think that this would be the place in your protocol to declare that. However, the gripper doesn't require ``load_instrument``! Whether your gripper requires a protocol is determined by the presence of :py:meth:`.ProtocolContext.move_labware` commands. See :ref:`moving-labware` for more details.
```
should go up in the Flex section

jwwojak replied:
Yikes. Yes it should. Fixing locally and will push.

ecormany commented on api/docs/v2/new_pipette.rst:
```diff
@@ -15,257 +15,211 @@ For information about liquid handling, see :ref:`v2-atomic-commands` and :ref:`v
+
 .. _new-multichannel-pipettes:
 
-Multi-Channel Pipettes
-======================
+Multi-Channel Pipettes and Well Plates
```
check the cross-references to make sure this won't add the text "and Well Plates" to links elsewhere.

jwwojak replied:
Let's go back to the original, "Multi-Channel Pipettes."  I didn't really like the extra I added on -- too long.

ecormany commented on api/docs/v2/new_pipette.rst:
```diff
@@ -15,257 +15,211 @@ For information about liquid handling, see :ref:`v2-atomic-commands` and :ref:`v
         load_name='opentrons_flex_96_tiprack_1000ul',
-        location='D1')
-        # Load a wellplate
+        location='D1')       
```
fix indents here, Python will consider this a syntax error

jwwojak replied:
2x yikes. Fixing locally. Will push revisions.

ecormany commented on api/docs/v2/new_pipette.rst:
```diff
@@ -15,257 +15,211 @@ For information about liquid handling, see :ref:`v2-atomic-commands` and :ref:`v
-        right.pick_up_tip(tiprack1['A1'])
+After loading our instruments and labware, let's tell the robot to pick up a pipette tip from location ``A1`` in ``tiprack1``::
+
+    right.pick_up_tip(tiprack1['A1'])
```
the code sets up tip tracking for this pipette. so i'd expect just `right.pick_up_tip()`. 

if you want to be explicit about the location on the tip rack, leave this line as is and remove the `tip_racks` parameter from `right = protocol.load_instrument(...`

jwwojak replied:
Hey, thanks for pointing that out. I had a TIL moment with this further down in the flow rate section. Let's go with `right.pick_up_tip()` because we've got that specified already.

ecormany commented on api/docs/v2/new_pipette.rst:
```diff
@@ -15,257 +15,211 @@ For information about liquid handling, see :ref:`v2-atomic-commands` and :ref:`v
-        # Dispense into column 3 of the plate with all 8 channels of the
-        # pipette at the top of their respective wells
-        right.dispense(volume=300, location=plate['A3'].top())
+After picking up a tip, let's tell the robot to aspirate 300 ul from the well plate at location ``A2``::
```
```suggestion
After picking up a tip, let's tell the robot to aspirate 300 µL from the well plate at location ``A2``::
```
unit style. there's a few of these to find and replace.

jwwojak replied:
Doing a search and replace now.

ecormany commented on api/docs/v2/new_pipette.rst:
```diff
@@ -15,257 +15,211 @@ For information about liquid handling, see :ref:`v2-atomic-commands` and :ref:`v
+        
+    right.aspirate(volume=300, location=plate['A2'])
+
+With the backmost pipette tip above location A2 on the well plate, all eight channels are above the eight wells in column 1.
```
```suggestion
With the backmost pipette tip above location A2 on the well plate, all eight channels are above the eight wells in column 2.
```

jwwojak replied:
Fixing locally. Will push.

ecormany commented on api/docs/v2/new_pipette.rst:
```diff
@@ -15,257 +15,211 @@ For information about liquid handling, see :ref:`v2-atomic-commands` and :ref:`v
+---------------------------------
 
-In general, you should specify wells in the first row of a well plate when using multi-channel pipettes. One exception to this rule is when using 384-well plates. The limited space between the wells in a 384-well plate and between the nozzles of a multi-channel pipette means the pipette accesses every other well in a column. Specifying well A1 accesses every other well starting with the first (ro ...
+In general, you should specify wells in the first row of a well plate when using multi-channel pipettes. An exception to this rule is when using 384-well plates. The greater well density means the nozzles of a multi-channel pipette can only accesses every other well in a column. Specifying well A1 accesses every other well starting with the first (rows A, C, E, G, I, K, M, and O). Similarly, spec ...
```
```suggestion
In general, you should specify wells in the first row of a well plate when using multi-channel pipettes. An exception to this rule is when using 384-well plates. The greater well density means the nozzles of a multi-channel pipette can only access every other well in a column. Specifying well A1 accesses every other well starting with the first (rows A, C, E, G, I, K, M, and O). Similarly, specifying well B1 also accesses every other well, but starts with the second (rows B, D, F, H, J, L, N, and P).
```
`can only access` or `only accesses`

jwwojak replied:
My god, I am illiterate. Fixing locally. Will commit.

ecormany commented on api/docs/v2/new_pipette.rst:
```diff
@@ -15,257 +15,211 @@ For information about liquid handling, see :ref:`v2-atomic-commands` and :ref:`v
-        +-------------------------+-----------+-------------------------+
-        | Flex 96-Channel Pipette | 5–1000 µL | ``flex_96channel_1000`` |
-        +-------------------------+-----------+-------------------------+
+    right.pick_up_tip(tiprack1['A1'])
```
make same changes to tip tracking as in 96-channel example

ecormany commented on api/docs/v2/new_pipette.rst:
```diff
@@ -15,257 +15,211 @@ For information about liquid handling, see :ref:`v2-atomic-commands` and :ref:`v
-        See the OT-2 Pipette Generations section below if you're using GEN1 pipettes on an OT-2. The GEN1 family includes the P10, P50, and P300 single- and multi-channel pipettes, along with the P1000 single-chanel model.
+    right.aspirate(volume=100, location=plate['A1'])
 
+Because of the limited clearance between wells, the eight pipette channels will only aspirate from every other well in the column (e.g. A1, C1, E1, G1, I1, K1, M1, O1).
```
```suggestion
The eight pipette channels will only aspirate from every other well in the column: A1, C1, E1, G1, I1, K1, M1, O1.
```
text above explains this well enough, i think

jwwojak replied:
Fixing locally. Will commit.

ecormany commented on api/docs/v2/new_pipette.rst:
```diff
@@ -15,257 +15,211 @@ For information about liquid handling, see :ref:`v2-atomic-commands` and :ref:`v
-      - P1000 Single GEN1 (100-1000 µL)
-
-The single- and multi-channel P50 GEN1 pipettes are the exceptions here. If your protocol uses a P50 GEN1 pipette, there is no backward compatibility with a related GEN2 pipette. To replace a P50 GEN1 with a corresponding GEN2 pipette, edit your protocol to load a P20 Single GEN2 (for volumes below 20 µL) or a P300 Single GEN2 (for volumes between 20 and 50 µL).
+Because of the limited clearance between wells, the eight pipette channels will only dispense into every other well in the column (e.g. B1, D1, F1, H1, J1, L1, N1, P1).
```
simplify in same way as line 179

jwwojak replied:
Fixing locally.

ecormany commented on api/docs/v2/new_pipette.rst:
```diff
@@ -15,257 +15,211 @@ For information about liquid handling, see :ref:`v2-atomic-commands` and :ref:`v
     def run(protocol: protocol_api.ProtocolContext):
     tiprack_left = protocol.load_labware(
-        load_name='opentrons_96_tiprack_300ul', location='1')
+        load_name='opentrons_flex_96_tiprack_200ul', location='1')
```
Flex labware, OT-2 deck slot. (nice that it doesn't matter!)

jwwojak replied:
Found 2 locations. Fixing slots 1 & 2 to D1 & D2.

jwwojak committed:
Revised flow rate code samples

Simplify examples. Consolidate each rate type into 1 sample block.

ecormany commented on api/docs/v2/new_pipette.rst:
```diff
@@ -15,257 +15,211 @@ For information about liquid handling, see :ref:`v2-atomic-commands` and :ref:`v
 
-    # You must specify the tip location for the left pipette, which was
-    # loaded without specifying tip_racks
+Next, let's specify the tip rack location for the left pipette, which was loaded without a ``tip_racks`` argument::
```
This explanation makes it sound like the `pick_up_tip` command will associate the tip rack with the pipette for the future, not just for that one action. Maybe something like `Let's pick up a tip with the left pipette. We need to specify the location as an argument of ``pick_up_tip()``, since we loaded the left pipette without a ``tip_racks`` argument.`

jwwojak replied:
Great recommendation. Fixing.

ecormany commented on api/docs/v2/new_pipette.rst:
```diff
@@ -15,257 +15,211 @@ For information about liquid handling, see :ref:`v2-atomic-commands` and :ref:`v
-    # And further calls to pick_up_tip will automatically progress through
-    # the tips in the rack
+Additional calls to ``pick_up_tip`` will automatically progress through the tips in the right rack::
+
```
the below code block is a great place to use inline comments to show the effect of a command. always put two spaces before the comment.

```
right_pipette.pick_up_tip()  # picks up from A2
right_pipette.drop_tip()
right_pipette.pick_up_tip()  # picks up from A3
right_pipette.drop_tip()
```

jwwojak replied:
Fixing locally. Taking note on comment spacing too.

ecormany commented on api/docs/v2/new_pipette.rst:
```diff
@@ -275,63 +229,124 @@ See also, :ref:`v2-atomic-commands` and :ref:`v2-complex-commands`.
+
+Let's tell the robot to aspirate, dispense, and blow out the liquid using default flow rates. Notice how you don't need to specify a ``flow_rate`` attribute to use the defaults::
+
+        pipette.aspirate(200, plate['A1'])
```
another good spot for comments
```
pipette.aspirate(200, plate['A1'])  # 160 µL/s
pipette.dispense(200, plate['A2'])  # 160 µL/s
pipette.blow_out()                  #  80 µL/s
```
here it makes sense to do some creative spacing with the last one for readability

jwwojak replied:
Fixing locally.

ecormany commented on api/docs/v2/new_pipette.rst:
```diff
@@ -275,63 +229,124 @@ See also, :ref:`v2-atomic-commands` and :ref:`v2-complex-commands`.
+        pipette.flow_rate.aspirate = 50
+        pipette.flow_rate.dispense = 100
+        pipette.flow_rate.blow_out = 75
+        pipette.aspirate(200, plate['A1'])
```
like above
```
pipette.aspirate(200, plate['A1'])  #  50 µL/s
pipette.dispense(200, plate['A2'])  # 100 µL/s
pipette.blow_out()                  #  75 µL/s
```

jwwojak replied:
Fixing locally.

ecormany commented on api/docs/v2/new_pipette.rst:
```diff
@@ -343,29 +358,53 @@ Flex Pipette Flow Rates
++-------------------------+-------------+-----------------+-----------------+-----------------+
+| Pipette Model           | Volume (µL) | Aspirate (µL/s) | Dispense (µL/s) | Blow-out (µL/s) |
++=========================+=============+=================+=================+=================+
+| Flex 1-Channel Pipette  | 1-50        | 8                                 | 4               |
```
wiiiiiide boi. let's list it twice: 8, 8, 4 and 160, 160, 80

jwwojak replied:
It is a wide load. We'll have to repeat the 160 and 80 three times because these have the flow rates for the 5-1000 uL pipettes:

- Flex 1-channel
- Flex 8-channel
- Flex 96-channel

With the revision, the only metrics that get the row spans are the volumes for the pipettes. Will fix and push.

ecormany commented on api/docs/v2/new_pipette.rst:
```diff
@@ -343,29 +358,53 @@ Flex Pipette Flow Rates
+| Flex 96-Channel Pipette |             |                                   |                 |
++-------------------------+-------------+-----------------------------------+-----------------+
+
+Additionally, all Flex pipettes have a well bottom clearance of 1mm for aspirate and dispense actions.
```
```suggestion
Additionally, all Flex pipettes have a well bottom clearance of 1 mm for aspirate and dispense actions.
```
😇

jwwojak replied:
Fixing locally. Will push.

ecormany commented on api/docs/v2/new_pipette.rst:
```diff
@@ -343,29 +358,53 @@ Flex Pipette Flow Rates
-+---------------------------------+-----------------------------------------------+
-
-.. what is head speed, is it important/significant? applies to gantry movement or pipette? need to find out.
+The following table provides data on the default aspirate, dispense, and blow-out flow rates (in µL/s) for OT-2 GEN2 pipettes. Because the flow rates are the same across types, we've consolidated the data into the "Flow Rates" column.
```
maybe `across all three actions` just so people don't think we're talking types of pipette.

jwwojak replied:
Yes, I thought this was wonk-a-doo weird. Will revise.

ecormany commented on api/docs/v2/new_pipette.rst:
```diff
@@ -343,29 +358,53 @@ Flex Pipette Flow Rates
+      - 20-300
+      - 94
+ 
 Additionally, all OT-2 GEN2 pipettes have a default head speed of 400 mm/second and a well bottom clearance of 1mm for aspirate and dispense actions.
```
`mm/s` and `1 mm` here…and then everything is neat and tidy!

jwwojak replied:
Fixing locally. Will push.

ecormany reviewed (changes requested):
Great, only a few things that didn't land in their final spot or needed a little cleanup. 🔜

jwwojak committed:
Update new_pipette.rst

Fix spacing

jwwojak committed:
Late day revisions

Changes from Ed's recent review. Once more unto the breech, friends!

jwwojak committed:
Fixing code

Used wrong name for pipette position argument.

jwwojak committed:
Update new_pipette.rst

Fixing code alignment.

jwwojak committed:
Update new_pipette.rst

Revision to section intro.

ecormany commented on api/docs/v2/new_pipette.rst:
```diff
@@ -6,436 +6,399 @@
-- Blow Out Default: 1000 µL/s
-- Minimum Volume: 100 µL
-- Maximum Volume: 1000 µL
+.. _pipette:
```
this is a very generic anchor name. it doesn't seem that we're linking here from anywhere, so we can remove it.

ecormany commented on api/docs/v2/new_pipette.rst:
```diff
@@ -6,436 +6,399 @@
     from opentrons import protocol_api
 
-    metadata = {'apiLevel': '|apiLevel|'}
+    metadata = {'apiLevel': |apiLevel|}
```
need the quotes here. the value has to be a string, or it raises an error.

ecormany commented on api/docs/v2/new_pipette.rst:
```diff
@@ -6,436 +6,399 @@
+
+To demonstrate these concepts, let's write a protocol that uses a Flex 8-Channel Pipette and a 96-well plate. We'll then aspirate and dispense a liquid to different locations on the same well plate. To start, let's load a pipette in the right mount and add our labware.
+
+.. code-block:: python
```
maybe include `requirements` here. the previous example is OT-2.

ecormany commented on api/docs/v2/new_pipette.rst:
```diff
@@ -6,436 +6,399 @@
+The ``load_instrument()`` method includes the optional argument ``tip_racks``. This parameter accepts a list of tip rack labware objects, which lets you to specify as many tip racks as you want. The advantage of using ``tip_racks`` is twofold. First, associating tip racks with your pipette allows for automatic tip tracking throughout your protocol. Second, it removes the need to specify tip locat ...
+        
+    def run(protocol: protocol_api.ProtocolContext):
+    tiprack_left = protocol.load_labware(
```
\+ indent level for everything under `def run`

ecormany committed:
Flex flow - just merge cells

ecormany committed:
fancier, more condensed table

ecormany committed:
code syntax fixes

ecormany reviewed (approved):
GTG. Massive improvement and lots of new info for Flex on this page. :shipit: 🦾

ecormany committed:
oops, :substitutions:

jwwojak commented:
This one was a _monster!_ Thanks, @ecormany!
